### ISSUE TITLE
EICrecon dependency on EPIC and IP6 repositories prevents inclusion in eic-shell environment

### ISSUE THREAD

**Issue by wdconinc** [eic/EICrecon#22]
Is the hard dependency on EPIC and IP6 necessary to build EICrecon? I am asking because we currently cannot include EICrecon in the eic-shell containers or benchmark pipelines (even for preliminary testing) because we treat the geometry descriptions as living on top of the infrastructure layers. In the past we have not had to know anything about the geometry descriptions to be able to compiler gaudi or juggler. Is that foreseen for EICrecon or will the EICrecon repository essentially only work for EPIC and IP6?

**Comment by wdconinc** [comment]
From what I can tell, you are linking to the detector libraries but not actually including any headers from the detectors nor do any symbols in the produced static libraries include references to the detector libraries.

**Comment by DraTeots** [comment]
I checked the code, not sure why EPIC and IP are referenced at all. Mybe David has more ideas on it.

**Comment by DraTeots** [comment]
I guess this is for this: 

```cmake
# The following ensures that on macosx the @rpath search paths go into the plugin
# itself and includes the plugin installation directory. This is needed so
# the ROOT dictionary libraries generated by PODIO are found.
# (n.b. there is probably a better way to do this since I believe this makes
# the plugin non-relocatable as it includes an absolute search path)
set_target_properties(dd4hep_plugin PROPERTIES SKIP_BUILD_RPATH FALSE)
set_target_properties(dd4hep_plugin PROPERTIES BUILD_WITH_INSTALL_RPATH TRUE)
set_target_properties(dd4hep_plugin PROPERTIES INSTALL_RPATH_USE_LINK_PATH TRUE)
set_property(
        TARGET dd4hep_plugin
        PROPERTY INSTALL_RPATH
        "${EIC_DD4HEP_HOME}/lib"
        "${IP6_DD4HEP_HOME}/lib"
)
```

**Comment by faustus123** [comment]
It was my understanding that the dd4hep plugin that gives access to the geometry to the rest of EICrecon does need to link to those libraries. How was this done in juggler/gaudi? Did you use libdl?

The real answer is that all of this is negotiable. I am not married to any of those details at the moment.

**Comment by wdconinc** [comment]
Everything juggler and gaudi know about any geometry from dd4hep is derived at run time. How it is done exactly is not really something any of us have had to look into in detail, so I can't say if it's libdl or otherwise.

**Comment by faustus123** [comment]
It does not appear to be something automatic. Without the libraries, it fails when parsing the XML as soon as it hits something it needs from the library. It does say something about a plugin that implies maybe there is a plugin mechanism in dd4hep we have not tapped into. Do you know if you needed any special environment variables to point to a dd4hep plugins directory?

**Comment by DraTeots** [comment]
Here it is
https://doc.athena-eic.org/en/latest/howto/full_simulation.html#how-xml-is-invoked-from-c-in-dd4hep

Before it was DD4HEP_LIBRARY_PATH and LD_LIBRARY_PATH
now they only check LD_LIBRARY_PATH

**Comment by wdconinc** [comment]
As far as I know it is just the LD_LIBRARY_PATH mechanism, but I think dd4hep also installs a component file.

**Comment by veprbl** [comment]
A user is supposed to source the [setup.sh](https://github.com/eic/epic/blob/main/templates/setup.sh.in) file to chose a certain geometry.

**Comment by faustus123** [comment]
So right now, on my system the setup.sh file complains that it cannot find ip6. This looks like it is because it is assuming the ip6 directory (or at least symlink) is installed in the share/epic directory. Maybe I'm just out of date on the build instructions for this? (my symlink to ip6 is in the top-level epic directory)

So assuming I fix that, does this mean in the dd4hep service we should be opening the file:

`${DETECTOR_PATH}/${DETECTOR}.xml`

And then does the dd4hep code automatically know to look at :

`/path/to/xmlfile/../../../lib64/lib${DETECTOR}.components`

to find the names of the libraries to search for in the LD_LIBRARY_PATH to dynamically open?

Seems a little convoluted, but I guess that is what it takes to push everything to run time so the executable for the dd4hep service is detector agnostic. Is there a reason not to place the components file right next to the XML? That would seem more logical IMHO. I guess so long as dd4hep is handling it, I don't really care. Just an observation.

side note: there needs to be a setup.csh so (t)csh users are not immediately thwarted.

**Comment by wdconinc** [comment]
> So assuming I fix that, does this mean in the dd4hep service we should be opening the file:

The reconstruction code should not assume any of this on compilation time, not an environment variable name, not a geometry location. There should be a way to specify on run-time which geometry a user wants to run on.

**Comment by wdconinc** [comment]
> side note: there needs to be a setup.csh so (t)csh users are not immediately thwarted.

See https://github.com/eic/ip6/pull/6 (and to a lesser extent https://github.com/eic/epic/pull/21) for the discussion.

**Comment by veprbl** [comment]
These are good points. Let me answer inline:

> So right now, on my system the setup.sh file complains that it cannot find ip6. This looks like it is because it is assuming the ip6 directory (or at least symlink) is installed in the share/epic directory. Maybe I'm just out of date on the build instructions for this? (my symlink to ip6 is in the top-level epic directory)

One is supposed to also source setup.sh from ip6 as well https://github.com/eic/ip6/blob/master/templates/setup.sh.in
(Also make sure to source $prefix/share/ip6/ip6/setup.sh and not $prefix/setup.sh as the latter may be deprecated)

> 
> So assuming I fix that, does this mean in the dd4hep service we should be opening the file:
> 
> `${DETECTOR_PATH}/${DETECTOR}.xml`

That's what juggler does.

> 
> And then does the dd4hep code automatically know to look at :
> 
> `/path/to/xmlfile/../../../lib64/lib${DETECTOR}.components`
> 
> to find the names of the libraries to search for in the LD_LIBRARY_PATH to dynamically open?

Apparently not. The names of detector components are found in xml detector type attribute (e.g. https://github.com/eic/epic/blob/2325756e887f27958f4901050d5c397371c1a6b4/compact/drich.xml#L55) and then DD4hep looks up all "$LD_LIBRARY_PATH/*.components" to pick the right library that provides such a detector type. This is also similar to how Gaudi finds its services.

> 
> Seems a little convoluted, but I guess that is what it takes to push everything to run time so the executable for the dd4hep service is detector agnostic. Is there a reason not to place the components file right next to the XML? That would seem more logical IMHO. I guess so long as dd4hep is handling it, I don't really care. Just an observation.

That is a question for the DD4hep upstream.

> 
> side note: there needs to be a setup.csh so (t)csh users are not immediately thwarted.

That is a known issue (e.g. https://github.com/eic/ip6/pull/6 and https://github.com/eic/epic/pull/21 was a generally good approach, but we got sidetracked)

**Comment by sly2j** [comment]
The workflow to support here is that if I am doing work on the detector, I should not need to worry about the reconstruction. For example, I might be working on adding detailed support frames to the detector plugin in a local working directory, while wanting to check the impact on a metric like calorimeter energy resolution. I may be working off a pre-compiled containerized environment. This requires no work on the reconstruction and and therefore the user should not have to care about this, except for telling DD4hep where to find the new plugins. This will also make quick CI checks faster, if we only need to worry about building our updated detector setup.

**Comment by wdconinc** [comment]
> Is there a reason not to place the components file right next to the XML?

The components file describes what's in the `lib<detector>.so` file right next to it. Essentially, the components file is what you get when you do `nm /opt/local/lib/libepic.so | grep epic_`. It is not related to the xml at all, which is the part that defines how you use those components and can vary independently.

**Comment by faustus123** [comment]
Right, I guess that is my point. If I have the libepic.so file I can just run nm on it to see what is in there. The XML file though is pretty useless without the components file.

**Comment by veprbl** [comment]
> Right, I guess that is my point. If I have the libepic.so file I can just run nm on it to see what is in there. The XML file though is pretty useless without the components file.

Detector description can come in the form of only the xml files that use some generic DD4hep detector descriptions from some meta library (e.g. https://github.com/AIDASoft/DD4hep/tree/master/DDDetectors/src).

**Comment by faustus123** [comment]
Regarding:

> The reconstruction code should not assume any of this on compilation time, not an environment variable name, not a geometry location. There should be a way to specify on run-time which geometry a user wants to run on.

Correct. The code looks for a configuration parameter. Well, it does look in the environment for a default, but if a config. parameter is given, that is used instead. No hard-coded paths are ever compiled in. I don't think it should be forbidden to fall back on an environment variable. It is required for the build and saves every user from having to add a ton of lines to their jana configuration that are specific to their build.

**Comment by wdconinc** [comment]
> I don't think it should be forbidden to fall back on an environment variable.

Environment variables are very difficult to handle from a preservation and reproducibility point of view since they are not manifest in the running configuration and cannot easily be archived.

In any case, back to the topic at hand, no environment variables that might be used at run time as fall-back defaults should be required at compile time.

**Comment by faustus123** [comment]
OK, things seem pretty clear now and I think there is general agreement. The dd4hep service needs to:

1. Not link the detector,ip6 libraries and let dd4hep dynamically load them at run time
2. Look to the DETECTOR and DETECTOR_PATH envars for defaults, but provide a config. parameter option which will take precedence

And users need an environment that at least has the LD_LIBRARY_PATH set correctly. Preferably by sourcing the appropriate setup.(c)sh files.

**Comment by faustus123** [comment]
> Environment variables are very difficult to handle from a preservation and reproducibility point of view since they are not manifest in the running configuration and cannot easily be archived.

In any case, back to the topic at hand, no environment variables that might be used at run time as fall-back defaults should be required at compile time.

Oh yes, maybe I misunderstood your earlier statement. The environment of the process at run time is used, not the environment at build time.

I think we are also in agreement that nothing should be left to the environment alone. This should only be used to seed the default value of a config parameter (at run time). The full configuration used can then be easily archived by telling JANA to write it to a file. This captures command line settings as well.

**Comment by DraTeots** [comment]
My point from Jana2 view. 

1. One has to remove any references of EPIC and IP6 in EICRecon 
2. dd4hep plugin should have xml parameter, which will be used like this: 

   ```bash
   eicrecon ... -cdd4hep:xml=${DETECTOR_PATH}/${DETECTOR}.xml
   # or
   eicrecon ... -cdd4hep:xml=/my/path/my_geometry.xml
   ```
   
   or in config/python/whatever
   
   ```
   dd4hep:xml = /path/to/file.xml
   ```

3. Today we discussed that it would be good if JANA took also an environment variables `${DETECTOR_PATH}/${DETECTOR}` to open xml by default. After thinking a bit I strongly believe it to be bad. The configuration of eicrecon goes through flags/config/python if several parameters are controlled by environment while the most of the others - don't it: 
    
    1. breaks consistency
    2. provides ways for implicit silent errors 
    3. confuses those who try to figure out and not finding a flag, while it exists in environment!
    4. kill kittens probably too
    
   My final arguments, are that 

   1. for Juggler those environment variables were a way to communicate with surroundings. For jana2 they are not. 
   2. For dd4hep we use file.xml explicitly as a flag in each application. And it is actually convenient. Or at least I haven't heard any complains about it.

**Comment by faustus123** [comment]
So I think we are just going to disagree on this one. I cannot see the logic of requiring the user to type in `-cdd4hep:xml=${DETECTOR_PATH}/${DETECTOR}.xml` every time they run the program which uses exactly the same environment variables we can access programatically. They will type exactly this >99% of the time. At least until they decide it is too much trouble and put it into a script. This is all just to avoid the appearance of using environment variables. I will point out that the ip6/epic geometry packages rely heavily on `LD_LIBRARY_PATH` to be set up for finding `.components` files and plugins it dynamically loads with libdl. You may call this "magic", but most users will consider this the "expected behavior".

We are definitely in agreement that the user should be able to set this on the command line or in a configuration file. I think here we see a good example of one of the main differences in how Gaudi  and JANA configure things. Gaudi mixes two concepts into their configuration files:

1. Algorithm connections and default settings
2. Deviations from the default settings (deltas)

This is evident in the >1k line configuration file Juggler uses. JANA's configuration is intended only for 2. so that it is much more obvious where you deviate from the standard configuration. Both systems allow full flexibility in configuring things at run time. JANA just pushes package authors to maintain their defaults in the package code rather than all authors trying to maintain them it the same global config. file.

So why is this relevant? Because if people use JANA as intended, they will have minimal command line arguments and tiny or even no configuration file. They will not be setting the geometry in their JANA config. file unless they intend to use a non-standard geometry. 

So to address your points:

> i. breaks consistency

The simulation program does not use JANA configuration files, but I assume pulls it's geometry location from the environment. Using the environment here will ensure consistency between running both the simulation and the reconstruction in the same session

I might also claim that environment variables are consistent with how dd4hep already uses LD_LIBRARY_PATH for plugins and .components files (i.e it does not use it for actual shared libraries).

>ii. provides multiple ways for terrible silent errors (when calculation goes OK, but it is not what was intended)

Yes, but the same is true for PATH and we're not asking people to type in the full path to `eicrecon` every time they run it.
The geometry file location used is printed to the screen and captured in the JANA configuration parameter manager when the job is run. The JANA configurations can be dumped to a text file at the end of the job. They will also be written to the podio output file so there will at least be a record of every setting used, regardless of where it came from. This will still be silent of course.

I personally have no faith that forcing this setting into a config file or onto the command line will have the affect you want. Users will treat it as a black box anyway, relegating the setting to a script or line in a config. file that they never pay attention to. There are simply too many parameters they will need to use to expect them to pay close attention to them all.

>iii. confuses those who try to figure out and not finding a flag. Flag exists in listing, is not provided and geometry magically opens.

I think maybe I addressed this above. I'm sure if I fully understand what is meant by "Flag exists in listing, is not provided...". Bottom line is, the value will be shown to the user and if they understand what it should be, they will know if it is an error or not. If they don't understand, they will be thrilled that it works since they would not know what to set it to otherwise.

> iv. kill kittens probably too

I claim we will save many more kittens with an envar fallback than without it.

Finally, anyone is perfectly free to pass this setting just as you describe. If they want that level of certainty, then that is their choice. "Choice" is the key word though. People may choose not set the value in their environment as well in which case it will fail with a useful error message unless they supply the config. parameter explicitly.

**Comment by wdconinc** [comment]
> Gaudi mixes two concepts into their configuration files:
> 1. Algorithm connections and default settings
> 2. Deviations from the default settings (deltas)

You could also say that JANA2 mixes two concepts in the source files:
1. Algorithm specifications
2. Algorithm connections

It is very common in our current benchmarks and in user code to run the same data streams simultaneously through different algorithms, or to run different data streams through the same algorithm. The JANA2 approach of having a `libBEMC.a` that includes "the" algorithm for the BEMC does not leave much flexibility to the user. It also forces duplication (and potential inconsistency) between `libBEMC.a` and the plugin for the backward Ecal which may use exactly the same clustering or digitization algorithm.

**Comment by faustus123** [comment]
I guess I don't understand your statement. JANA allows you to configure to use algorithm "A" or "B", but just defaults to "A" if you don't specify. How is that less flexible?

Also, JANA does not force the use of libBEMC.a. It is an option that is possible in the flexible design choices JANA allows. The current system we are playing with for EICrecon produces both libBEMC.a and BEMC.so. I've been considering whether the eicrecon program should just use the plugins rather than the statically linked libraries. Again, implementation choice, not a JANA requirement.

**Comment by faustus123** [comment]
>You could also say that JANA2 mixes two concepts in the source files:
>
>Algorithm specifications
>Algorithm connections

BTW, this would mean Gaudi mixes 3 concepts.

**Comment by wdconinc** [comment]
> BTW, this would mean Gaudi mixes 3 concepts.

Gaudi can provide solutions to these 3 concepts in different aspects of its framework. We're discussing how to keep the 3 concepts intentionally separate (and, maybe, you are suggesting they should be mixed anyway). The question is really whether for any of these concepts is there 1 place where it is specified, or is it a combination of places that can change independently. No one was saying that we need a different piece of software for each of the 3 concepts.

Gaudi algorithms source files do not specify connections between themselves in the algorithm code, just the input and output data types. The configuration files are the only place where connections are specified.

Gaudi algorithms do specify default settings, and configurations specify deltas. I have to admit that I think this violates separation, and I would be more comfortable with specifying *everything* in configuration files, in order to be manifest. Changing defaults results in changing results when those parameters are implicitly relied upon (sometimes while other defaults are not relied upon).

It doesn't really matter what Gaudi does, since we are not using it, but it matters what JANA2 and EICrecon do. In that context it makes sense to think about the assumptions that you have, and how other systems have evolved to reflect the needs of users. That includes Gaudi.

**Comment by DraTeots** [comment]
Several technical misconceptions are made here. Let me clarify them (as I was the one who poured oil into a fire). I will try NOT to compare JANA2 with Gaudy as much as possible as my goal to clear on technical details. 

> The JANA2 approach of having a libBEMC.a that includes "the" algorithm for the BEMC does not leave much flexibility to the user.

libBEMC is actually a flexible way to provide _a_ default algorithm for BEMC. Don't get confused by a library. If BEMC data is not needed nothing is called at all. But the most important, if user explicitly specify a plugin with factory replacing the default BEMC factory from libBEMC it will be replaced. It will not be in silence as JANA do both: reports of what exactly is being used and can produce a definitive configuration output. 

>  for the BEMC does not leave much flexibility to the user.

Even more, using factory tags one gets the full flexibility to overwrite a default algorithms or use A and B in parallel, etc. That is actually good, if you take best practices of making large libraries such as standard library for languages, etc - this approach is usually chosen and works pretty well across decades. I can summarize it: if you try it - it just runs. If you need advanced features - components might be replaced to whatever you wants. Here we come to the next important point:

> We're discussing how to keep the 3 concepts intentionally separate 

It is very important. When you try to develop a modular software with loose coupling of parts,  you have to choose based on which aspects you do the separation. And those aspects are a bit different for Gaudi and Jana2. And some things that are important for system one - are less important for the system two. Jana does automatic linking between algorithms and it is one of its advantages. EICrecon is now being also oriented towards a case that different group of experts can work on their own systems. It is not exactly JANA2 requirement but consider it is to be a best practices.  That is very important to consider for the next question: 

> ...  to run the same data streams simultaneously through different algorithms, or to run different data streams through the same algorithm. ...  It also forces duplication (and potential inconsistency) between libBEMC.a and the plugin for the backward Ecal which may use exactly the same clustering or digitization algorithm.

TL;DR; there is no problem to have one algorithm and pipe different data types through it in JANA.

To answer it more thoroughly, why BEMC.so /a exists at all? In our software case each system is an unique piece of harware/electronics and software around it. There is an unique software parts for each detector (almost all the time) and higher level algorithms which can be shared between the detectors. Where those detector specific parts should be? For that reason BEMC directory is created and not to put everything for BEMC including clustering algorithms in that directory. We are actively putting things together and actually used it as some code example showing one of the approaches. What about duplication? There are so many architectural patterns not to duplicate your code. There are tools in JANA2. There are tools in C++. We certainly will not copy/paste clustering in each subsystem specific plugin. 

> I would be more comfortable with specifying everything in configuration files, in order to be manifest. 

It is tricky as there are more things. At least alignment, calibrations and conditions. Which change in time and are silent state changers. IDK what we will use, but in theory all of them allow to use some "fixed" versions of provided data. But! in reality the problem of creating a manifest manually becomes too multidimensional and much more mutable in time (groups constantly redoing and improving calibrations, etc). Again the approaches of frameworks are different, automatically solving one of the aspects seems good actually.

**Comment by DraTeots** [comment]
@faustus123 can't write another long text. You convinced me. I was wrong. 

P.S. Lets go save more kittens

**Comment by wdconinc** [comment]
Implemented in https://github.com/eic/EICrecon/commit/42b6ce049dffab8369e12d9c002ef4c426758421
